framework/src/loops/ComputeLinearFVElementalThread.C

```diff
@@ -0,0 +1,134 @@
+
+ComputeLinearFVElementalThread::ComputeLinearFVElementalThread(
+    FEProblemBase & fe_problem,
+    const unsigned int linear_system_num,
+    const Moose::FV::LinearFVComputationMode mode,
+    const std::set<TagID> & tags)
+  : _fe_problem(fe_problem),
+    _linear_system_number(linear_system_num),
+    _mode(mode),
+    _tags(tags),
+    _subdomain(Moose::INVALID_BLOCK_ID),
+    _old_subdomain(Moose::INVALID_BLOCK_ID)
+{
+}
+
+// Splitting Constructor
+ComputeLinearFVElementalThread::ComputeLinearFVElementalThread(ComputeLinearFVElementalThread & x,
+                                                               Threads::split /*split*/)
+  : _fe_problem(x._fe_problem),
+    _linear_system_number(x._linear_system_number),
+    _mode(x._mode),
+    _tags(x._tags),
+    _subdomain(x._subdomain),
+    _old_subdomain(x._old_subdomain)
+{
+}
+
+void
+ComputeLinearFVElementalThread::operator()(const ElemInfoRange & range)
+{
+  ParallelUniqueId puid;
+  _tid = puid.id;
+
+  _old_subdomain = Moose::INVALID_BLOCK_ID;
+
+  printGeneralExecutionInformation();
+
+  // Iterate over all the elements in the range
+  for (const auto & elem_info : range)
+  {
+    _subdomain = elem_info->subdomain_id();
+    if (_subdomain != _old_subdomain)
+    {
+      fetchSystemContributionObjects();
+      printBlockExecutionInformation();
+    }
+
+    const Real elem_volume = elem_info->volume() * elem_info->coordFactor();
+    for (auto kernel : _fv_kernels)
+    {
+      kernel->setCurrentElemInfo(elem_info);
+      kernel->setCurrentElemVolume(elem_volume);
+      if (_mode == Moose::FV::LinearFVComputationMode::Matrix ||
+          _mode == Moose::FV::LinearFVComputationMode::FullSystem)
+        kernel->addMatrixContribution();
+      if (_mode == Moose::FV::LinearFVComputationMode::RHS ||
+          _mode == Moose::FV::LinearFVComputationMode::FullSystem)
+        kernel->addRightHandSideContribution();
+    }
+  }
+}
+
+void
+ComputeLinearFVElementalThread::join(const ComputeLinearFVElementalThread & /*y*/)
+{
+}
+
+void
+ComputeLinearFVElementalThread::fetchSystemContributionObjects()
+{
+  const auto system_number = _fe_problem.getLinearSystem(_linear_system_number).number();
+  _fv_kernels.clear();
+  std::vector<LinearFVElementalKernel *> kernels;
+  auto base_query = _fe_problem.theWarehouse()
+                        .query()
+                        .template condition<AttribSysNum>(system_number)
+                        .template condition<AttribSubdomains>(_subdomain)
+                        .template condition<AttribSystem>("LinearFVElementalKernel")
+                        .template condition<AttribThread>(_tid);
+
+  if (_mode == Moose::FV::LinearFVComputationMode::Matrix)
+    base_query.condition<AttribMatrixTags>(_tags).queryInto(kernels);
+  else
+    base_query.condition<AttribVectorTags>(_tags).queryInto(kernels);
+  _old_subdomain = _subdomain;
+
+  _fv_kernels = std::set<LinearFVElementalKernel *>(kernels.begin(), kernels.end());
```

Comment:
dont really see why you keep a set instead of a vector?
if you want to be ultra optimized, vector is better